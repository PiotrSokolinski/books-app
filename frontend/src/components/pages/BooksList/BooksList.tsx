import {
  Datagrid,
  List,
  useResourceContext,
  TextField,
  NumberField,
  useTranslate,
  SearchInput,
  TextInput,
} from 'react-admin';

import { Empty } from '../../molecules/Empty/Empty';
import { Label } from '../../atoms/Label/Label';
import { ListActions } from '../../organisms/ListActions/ListActions';

export const BooksList = () => {
  const resource = useResourceContext();
  const translate = useTranslate();

  const filters = [
    <SearchInput
      name="search"
      source="q"
      alwaysOn
      placeholder={translate('booksList.search')}
    />,
    <TextInput
      name="isbnNumber"
      label={translate('booksList.isbnNumber')}
      source="isbnNumber"
    />,
    <TextInput
      name="rating"
      label={translate('booksList.rating')}
      source="rating"
    />,
  ];
  return (
    <List
      exporter={false}
      resource={resource}
      title={translate('booksList.barTitle')}
      actions={<ListActions />}
      filters={filters}
      filterDefaultValues={{ search: '', isbnNumber: '', rating: '' }}
      sort={{ field: 'isbnNumber', order: 'ASC' }}
      debounce={500}
      empty={false}
    >
      <Datagrid empty={<Empty resource={resource} />}>
        <TextField source="id" label={<Label textId="booksList.id" />} />
        <TextField source="title" label={<Label textId="booksList.title" />} />
        <TextField
          source="author"
          label={<Label textId="booksList.author" />}
        />
        <TextField
          source="isbnNumber"
          label={<Label textId="booksList.isbnNumber" />}
        />
        <NumberField
          source="numberOfPages"
          label={<Label textId="booksList.numberOfPages" />}
        />
        <NumberField
          source="rating"
          label={<Label textId="booksList.rating" />}
        />
      </Datagrid>
    </List>
  );
};
