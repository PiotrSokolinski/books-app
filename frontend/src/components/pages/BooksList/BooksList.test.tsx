import {
  act,
  fireEvent,
  render,
  screen,
  waitFor,
} from '@testing-library/react';
import { describe, it, expect } from 'vitest';
import { DataProvider, ResourceContextProvider } from 'react-admin';

import { BooksList } from './BooksList';
import { AdminContextDecorator } from '../../../utils/AdminContextDecorator';

describe('BooksList', () => {
  it('renders the BooksList component with correct filters and datagrid', async () => {
    await act(async () =>
      render(
        <AdminContextDecorator>
          <ResourceContextProvider value="books">
            <BooksList />
          </ResourceContextProvider>
        </AdminContextDecorator>,
      ),
    );

    await waitFor(() => {
      expect(screen.getByPlaceholderText('Search')).toBeInTheDocument();
      expect(screen.getByText('ID')).toBeInTheDocument();
      expect(screen.getByText('Title')).toBeInTheDocument();
      expect(screen.getByText('Author')).toBeInTheDocument();
      expect(screen.getByText('ISBN number')).toBeInTheDocument();
      expect(screen.getByText('Number of pages')).toBeInTheDocument();
      expect(screen.getByText('Rating')).toBeInTheDocument();
      expect(screen.getByText('Create')).toBeInTheDocument();
      expect(screen.getByText('Add filter')).toBeInTheDocument();
    });
  });

  it('renders the BooksList component and can click create button', async () => {
    await act(async () =>
      render(
        <AdminContextDecorator>
          <ResourceContextProvider value="books">
            <BooksList />
          </ResourceContextProvider>
        </AdminContextDecorator>,
      ),
    );

    await waitFor(() => {
      const createButton = screen.getByText('Create');
      expect(createButton).toBeInTheDocument();
      fireEvent.click(createButton);
    });
  });

  it('renders the BooksList component and can click row', async () => {
    await act(async () =>
      render(
        <AdminContextDecorator>
          <ResourceContextProvider value="books">
            <BooksList />
          </ResourceContextProvider>
        </AdminContextDecorator>,
      ),
    );

    await waitFor(() => {
      const id = screen.getByText('1');
      expect(id).toBeInTheDocument();
      fireEvent.click(id);
    });
  });

  it('renders the ListActions and Empty components correctly', async () => {
    const dataProvider: DataProvider = {
      getList: Promise.resolve({
        total: 0,
        data: [],
      }),
    } as unknown as DataProvider;
    await act(async () =>
      render(
        <AdminContextDecorator dataProvider={dataProvider}>
          <ResourceContextProvider value="books">
            <BooksList />
          </ResourceContextProvider>
        </AdminContextDecorator>,
      ),
    );
    await waitFor(() => {
      expect(screen.getByText('No Books yet.')).toBeInTheDocument();
    });
  });

  it('matches the snapshot', async () => {
    let asFragment: () => DocumentFragment = () =>
      document.createDocumentFragment();
    await act(async () => {
      const result = render(
        <AdminContextDecorator>
          <ResourceContextProvider value="books">
            <BooksList />
          </ResourceContextProvider>
        </AdminContextDecorator>,
      );
      asFragment = result.asFragment;
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
