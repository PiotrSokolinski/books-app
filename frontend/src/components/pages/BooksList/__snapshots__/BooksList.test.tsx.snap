// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BooksList > matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="list-page css-1ycqwd-RaList-root"
  >
    <div
      class="RaList-main"
    >
      <div
        class="RaList-actions css-7k4ndg-RaListToolbar-root"
      >
        <form
          class="css-18hcsan-RaFilterForm-root"
        >
          <div
            class="filter-field RaFilterForm-filterFormInput css-knw8zc-RaFilterFormInput-root"
            data-source="q"
          >
            <div
              class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root ra-input ra-input-q css-1usige6-MuiFormControl-root-MuiTextField-root-RaResettableTextField-root-RaSearchInput-root"
            >
              <div
                class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd MuiInputBase-hiddenLabel css-te332i-MuiInputBase-root-MuiFilledInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputHiddenLabel MuiInputBase-inputAdornedEnd css-fe7vbu-MuiInputBase-input-MuiFilledInput-input"
                  id=":r1b:"
                  name="search"
                  placeholder="Search"
                  type="text"
                  value=""
                />
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-filled MuiInputAdornment-hiddenLabel MuiInputAdornment-sizeSmall css-jv61v3-MuiInputAdornment-root"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-colorDisabled MuiSvgIcon-fontSizeMedium css-1db085k-MuiSvgIcon-root"
                    data-testid="SearchIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <div
              class="RaFilterFormInput-spacer"
            >
              Â 
            </div>
          </div>
          <div
            class="RaFilterForm-clearFix"
          />
        </form>
        <span />
        <div
          class="MuiToolbar-root MuiToolbar-dense css-8czsxw-MuiToolbar-root-RaTopToolbar-root"
        >
          <div
            class="css-t6tio9-RaFilterButton-root"
          >
            <button
              aria-haspopup="true"
              aria-label="Add filter"
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary add-filter css-5g6lod-MuiButtonBase-root-MuiButton-root-RaButton-root"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="FilterListIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10 18h4v-2h-4zM3 6v2h18V6zm3 7h12v-2H6z"
                  />
                </svg>
              </span>
              Add filter
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <a
            aria-label="Create"
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary RaCreateButton-root css-f5d03y-MuiButtonBase-root-MuiButton-root-RaButton-root-RaCreateButton-root"
            href="#/books/create"
            tabindex="0"
          >
            <span
              class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="AddIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6z"
                />
              </svg>
            </span>
            Create
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </a>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root RaList-content css-bhp9pd-MuiPaper-root-MuiCard-root"
      >
        <div
          class="RaDatagrid-root css-14mup8f-RaDatagrid-root"
        >
          <div
            class="css-19tabqp-RaBulkActionsToolbar-root"
          >
            <div
              class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular RaBulkActionsToolbar-toolbar RaBulkActionsToolbar-collapsed css-hyum1k-MuiToolbar-root"
              data-test="bulk-actions-toolbar"
            >
              <div
                class="RaBulkActionsToolbar-title"
              >
                <button
                  aria-label="Unselect"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall RaBulkActionsToolbar-icon css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  title="Unselect"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                    data-testid="CloseIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
                <h6
                  class="MuiTypography-root MuiTypography-subtitle1 css-1yqpi80-MuiTypography-root"
                >
                  0 items selected
                </h6>
              </div>
              <div
                class="MuiToolbar-root MuiToolbar-dense RaBulkActionsToolbar-topToolbar css-8czsxw-MuiToolbar-root-RaTopToolbar-root"
              >
                <button
                  aria-label="Delete"
                  class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary css-1xlp554-MuiButtonBase-root-MuiButton-root-RaButton-root-RaBulkDeleteWithUndoButton-root"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="DeleteIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
                      />
                    </svg>
                  </span>
                  Delete
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
          <div
            class="RaDatagrid-tableWrapper"
          >
            <table
              class="MuiTable-root RaDatagrid-table css-rqglhn-MuiTable-root"
            >
              <thead
                class="MuiTableHead-root RaDatagrid-thead css-15wwp11-MuiTableHead-root"
              >
                <tr
                  class="MuiTableRow-root MuiTableRow-head RaDatagrid-row RaDatagrid-headerRow css-1q1u3t4-MuiTableRow-root"
                >
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-paddingCheckbox MuiTableCell-sizeSmall RaDatagrid-headerCell css-blmy2t-MuiTableCell-root"
                    scope="col"
                  >
                    <span
                      class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium select-all css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
                    >
                      <input
                        aria-label="Select all"
                        class="PrivateSwitchBase-input css-1m9pwf3"
                        data-indeterminate="false"
                        type="checkbox"
                      />
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="CheckBoxOutlineBlankIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </span>
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall RaDatagrid-headerCell column-id css-kp6q86-MuiTableCell-root-RaDatagridHeaderCell-root"
                    scope="col"
                  >
                    <span
                      aria-label="Sort by id ascending"
                      class="MuiButtonBase-root MuiTableSortLabel-root css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
                      data-field="id"
                      data-mui-internal-clone-element="true"
                      data-order="ASC"
                      role="button"
                      tabindex="0"
                    >
                      <h6
                        class="MuiTypography-root MuiTypography-h6 css-snu5bf-MuiTypography-root"
                      >
                        ID
                      </h6>
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon RaDatagridHeaderCell-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
                        data-testid="ArrowDownwardIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                        />
                      </svg>
                    </span>
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall RaDatagrid-headerCell column-title css-kp6q86-MuiTableCell-root-RaDatagridHeaderCell-root"
                    scope="col"
                  >
                    <span
                      aria-label="Sort by title ascending"
                      class="MuiButtonBase-root MuiTableSortLabel-root css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
                      data-field="title"
                      data-mui-internal-clone-element="true"
                      data-order="ASC"
                      role="button"
                      tabindex="0"
                    >
                      <h6
                        class="MuiTypography-root MuiTypography-h6 css-snu5bf-MuiTypography-root"
                      >
                        Title
                      </h6>
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon RaDatagridHeaderCell-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
                        data-testid="ArrowDownwardIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                        />
                      </svg>
                    </span>
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall RaDatagrid-headerCell column-author css-kp6q86-MuiTableCell-root-RaDatagridHeaderCell-root"
                    scope="col"
                  >
                    <span
                      aria-label="Sort by author ascending"
                      class="MuiButtonBase-root MuiTableSortLabel-root css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
                      data-field="author"
                      data-mui-internal-clone-element="true"
                      data-order="ASC"
                      role="button"
                      tabindex="0"
                    >
                      <h6
                        class="MuiTypography-root MuiTypography-h6 css-snu5bf-MuiTypography-root"
                      >
                        Author
                      </h6>
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon RaDatagridHeaderCell-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
                        data-testid="ArrowDownwardIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                        />
                      </svg>
                    </span>
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall RaDatagrid-headerCell column-isbnNumber css-kp6q86-MuiTableCell-root-RaDatagridHeaderCell-root"
                    scope="col"
                  >
                    <span
                      aria-label="Sort by isbn number descending"
                      class="MuiButtonBase-root MuiTableSortLabel-root Mui-active css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
                      data-field="isbnNumber"
                      data-mui-internal-clone-element="true"
                      data-order="ASC"
                      role="button"
                      tabindex="0"
                    >
                      <h6
                        class="MuiTypography-root MuiTypography-h6 css-snu5bf-MuiTypography-root"
                      >
                        ISBN number
                      </h6>
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon RaDatagridHeaderCell-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
                        data-testid="ArrowDownwardIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                        />
                      </svg>
                    </span>
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall RaDatagrid-headerCell column-numberOfPages css-25u9ct-MuiTableCell-root-RaDatagridHeaderCell-root"
                    scope="col"
                  >
                    <span
                      aria-label="Sort by number of pages ascending"
                      class="MuiButtonBase-root MuiTableSortLabel-root css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
                      data-field="numberOfPages"
                      data-mui-internal-clone-element="true"
                      data-order="ASC"
                      role="button"
                      tabindex="0"
                    >
                      <h6
                        class="MuiTypography-root MuiTypography-h6 css-snu5bf-MuiTypography-root"
                      >
                        Number of pages
                      </h6>
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon RaDatagridHeaderCell-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
                        data-testid="ArrowDownwardIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                        />
                      </svg>
                    </span>
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall RaDatagrid-headerCell column-rating css-25u9ct-MuiTableCell-root-RaDatagridHeaderCell-root"
                    scope="col"
                  >
                    <span
                      aria-label="Sort by rating ascending"
                      class="MuiButtonBase-root MuiTableSortLabel-root css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
                      data-field="rating"
                      data-mui-internal-clone-element="true"
                      data-order="ASC"
                      role="button"
                      tabindex="0"
                    >
                      <h6
                        class="MuiTypography-root MuiTypography-h6 css-snu5bf-MuiTypography-root"
                      >
                        Rating
                      </h6>
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon RaDatagridHeaderCell-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
                        data-testid="ArrowDownwardIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                        />
                      </svg>
                    </span>
                  </th>
                </tr>
              </thead>
              <tbody
                class="MuiTableBody-root datagrid-body RaDatagrid-tbody css-apqrd9-MuiTableBody-root"
              >
                <tr
                  class="MuiTableRow-root MuiTableRow-hover RaDatagrid-row RaDatagrid-rowEven RaDatagrid-selectable css-1q1u3t4-MuiTableRow-root"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingCheckbox MuiTableCell-sizeSmall css-2u8re1-MuiTableCell-root"
                  >
                    <span
                      aria-label="Select this row"
                      class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium select-item RaDatagrid-checkbox css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
                    >
                      <input
                        class="PrivateSwitchBase-input css-1m9pwf3"
                        data-indeterminate="false"
                        type="checkbox"
                      />
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="CheckBoxOutlineBlankIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </span>
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall column-id RaDatagrid-rowCell css-dsuxgy-MuiTableCell-root"
                  >
                    <span
                      class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
                    >
                      1
                    </span>
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall column-title RaDatagrid-rowCell css-dsuxgy-MuiTableCell-root"
                  >
                    <span
                      class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
                    >
                      The Hobbit
                    </span>
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall column-author RaDatagrid-rowCell css-dsuxgy-MuiTableCell-root"
                  >
                    <span
                      class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
                    >
                      J.R.R. Tolkien
                    </span>
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall column-isbnNumber RaDatagrid-rowCell css-dsuxgy-MuiTableCell-root"
                  >
                    <span
                      class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
                    >
                      1234567890
                    </span>
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall column-numberOfPages RaDatagrid-rowCell css-dv0mq-MuiTableCell-root"
                  >
                    <span
                      class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
                    >
                      300
                    </span>
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall column-rating RaDatagrid-rowCell css-dv0mq-MuiTableCell-root"
                  >
                    <span
                      class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
                    >
                      4
                    </span>
                  </td>
                </tr>
                <tr
                  class="MuiTableRow-root MuiTableRow-hover RaDatagrid-row RaDatagrid-rowOdd RaDatagrid-selectable css-1q1u3t4-MuiTableRow-root"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingCheckbox MuiTableCell-sizeSmall css-2u8re1-MuiTableCell-root"
                  >
                    <span
                      aria-label="Select this row"
                      class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium select-item RaDatagrid-checkbox css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
                    >
                      <input
                        class="PrivateSwitchBase-input css-1m9pwf3"
                        data-indeterminate="false"
                        type="checkbox"
                      />
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="CheckBoxOutlineBlankIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </span>
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall column-id RaDatagrid-rowCell css-dsuxgy-MuiTableCell-root"
                  >
                    <span
                      class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
                    >
                      2
                    </span>
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall column-title RaDatagrid-rowCell css-dsuxgy-MuiTableCell-root"
                  >
                    <span
                      class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
                    >
                      The Lord of the Rings
                    </span>
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall column-author RaDatagrid-rowCell css-dsuxgy-MuiTableCell-root"
                  >
                    <span
                      class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
                    >
                      J.R.R. Tolkien
                    </span>
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall column-isbnNumber RaDatagrid-rowCell css-dsuxgy-MuiTableCell-root"
                  >
                    <span
                      class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
                    >
                      0987654321
                    </span>
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall column-numberOfPages RaDatagrid-rowCell css-dv0mq-MuiTableCell-root"
                  >
                    <span
                      class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
                    >
                      1,000
                    </span>
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall column-rating RaDatagrid-rowCell css-dv0mq-MuiTableCell-root"
                  >
                    <span
                      class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
                    >
                      5
                    </span>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
      <span
        class="MuiTablePagination-root css-jtlhu6-MuiTablePagination-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-78c6dr-MuiToolbar-root-MuiTablePagination-toolbar"
        >
          <div
            class="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
          />
          <p
            class="MuiTablePagination-selectLabel css-pdct74-MuiTablePagination-selectLabel"
            id=":r1m:"
          >
            Rows per page:
          </p>
          <div
            class="MuiInputBase-root MuiInputBase-colorPrimary MuiTablePagination-input css-16c50h-MuiInputBase-root-MuiTablePagination-select"
          >
            <div
              aria-controls=":r1n:"
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby=":r1m: :r1l:"
              class="MuiSelect-select MuiTablePagination-select MuiSelect-standard MuiInputBase-input css-194a1fa-MuiSelect-select-MuiInputBase-input"
              id=":r1l:"
              role="combobox"
              tabindex="0"
            >
              10
            </div>
            <input
              aria-hidden="true"
              aria-invalid="false"
              class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              tabindex="-1"
              value="10"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiTablePagination-selectIcon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
          <p
            class="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
          >
            1-2 of 2
          </p>
          <div
            class="MuiTablePagination-actions css-e8518-RaPaginationActions-root"
          />
        </div>
      </span>
    </div>
  </div>
</DocumentFragment>
`;
